c++/src/kj/compat/http.c++

```diff
@@ -1110,6 +1112,150 @@ kj::String HttpHeaders::toString() const {
+  if (*(++p) != 'y' && *p != 'Y') return false;
+  if (*(++p) != 't' && *p != 'T') return false;
+  if (*(++p) != 'e' && *p != 'E') return false;
+  if (*(++p) != 's' && *p != 'S') return false;
+  ++p;
+
+  p = skipSpace(p);
+  ptr = p;
+  return true;
+}
+
+enum HttpRangeParseError {
+  INVALID, // Parsing failed
+  UNSATISFIABLE // Able to parse and pointer advanced, incorrect for another reason
+};
+
+typedef kj::OneOf<HttpByteRange, HttpRangeParseError> HttpRangeParseResult;
+
+static HttpRangeParseResult consumeIntRange(const char*& ptr, uint64_t contentLength) {
+  const char* p = ptr;
+  p = skipSpace(p);
+  uint firstPos;
+  KJ_IF_MAYBE(n, consumeNumber(p)) {
+    firstPos = *n;
+  } else {
+    return INVALID;
+  }
+  p = skipSpace(p);
+  if (*(p++) != '-') return INVALID;
+  p = skipSpace(p);
+  auto maybeLastPos = consumeNumber(p);
+  p = skipSpace(p);
+
+  if (firstPos >= contentLength) {
+    ptr = p;
+    // "a valid bytes range-spec is satisfiable if it is ... an int-range with a first-pos that is
+    // less than the current length of the selected representation"
+    return UNSATISFIABLE;
+  }
+  KJ_IF_MAYBE(lastPos, maybeLastPos) {
+    // "An int-range is invalid if the last-pos value is present and less than the first-pos"
+    if (firstPos > *lastPos) return INVALID;
+    // "if the value is greater than or equal to the current length of the representation data
+    // ... interpreted as the remainder of the representation"
+    if (*lastPos >= contentLength) *lastPos = contentLength - 1;
+    ptr = p;
+    return HttpByteRange { firstPos, *lastPos };
+  } else {
+    // "if the last-pos value is absent ... interpreted as the remainder of the representation"
+    ptr = p;
+    return HttpByteRange { firstPos, contentLength - 1 };
+  }
+}
+
+static HttpRangeParseResult consumeSuffixRange(const char*& ptr, uint64_t contentLength) {
+  const char* p = ptr;
+  p = skipSpace(p);
+  if (*(p++) != '-') return INVALID;
+  p = skipSpace(p);
+  uint suffixLength;
+  KJ_IF_MAYBE(n, consumeNumber(p)) {
+    suffixLength = *n;
+  } else {
+    return INVALID;
+  }
+  p = skipSpace(p);
+
+  ptr = p;
+  if (suffixLength == 0) {
+    // "a valid bytes range-spec is satisfiable if it is ... a suffix-range with a non-zero
+    // suffix-length"
+    return UNSATISFIABLE;
+  } else if (suffixLength >= contentLength) {
+    // "if the selected representation is shorter than the specified suffix-length, the entire
+    // representation is used"
+    return HttpByteRange { 0, contentLength - 1 };
+  } else {
+    return HttpByteRange { contentLength - suffixLength, contentLength - 1 };
+  }
+}
+
+static HttpRangeParseResult consumeRangeSpec(const char*& ptr, uint64_t contentLength) {
+  KJ_SWITCH_ONEOF(consumeIntRange(ptr, contentLength)) {
+    KJ_CASE_ONEOF(range, HttpByteRange) {
+      return range;
+    }
+    KJ_CASE_ONEOF(error, HttpRangeParseError) {
+      if (error == UNSATISFIABLE) return error;
```

Comment:
Do we really want to go on parsing if `consumeIntRange()` reported INVALID? Maybe add comments explaining why, if that's the case?